azext_iot/tests/helpers.py

```diff
@@ -156,6 +159,127 @@ def get_agent_public_ip():
     return requests.head("https://www.wikipedia.org").headers["X-Client-IP"]
 
 
+def get_role_assignments(
+    scope: str,
+    assignee: str = None,
+    role: str = None,
+) -> json:
+    """
+    Get rbac permissions of resource.
+    """
+    role_flag = ""
+    assignee_flag = ""
+
+    if role:
+        role_flag = '--role "{}"'.format(role)
+
+    if assignee:
+        assignee_flag = '--assignee "{}"'.format(assignee)
+
+    return cli.invoke(
+        f'role assignment list --scope "{scope}" {role_flag} {assignee_flag}'
+    ).as_json()
+
+
+def assign_role_assignment(
+    role: str,
+    scope: str,
+    assignee: str,
+    max_tries=10,
+    wait=10
+):
+    """
+    Assign rbac permissions to resource.
+    """
+    output = None
+    tries = 0
+    while tries < max_tries:
+        role_assignments = get_role_assignments(scope=scope, role=role)
+        role_assignment_principal_ids = [assignment["principalId"] for assignment in role_assignments]
+        if assignee in role_assignment_principal_ids:
+            break
+        # else assign role to scope and check again
+        output = cli.invoke(
+            f'role assignment create --assignee "{assignee}" --role "{role}" --scope "{scope}"'
+        )
+        sleep(wait)
+        tries += 1
+    return output
+
+
+def delete_role_assignment(
+    scope: str,
+    assignee: str,
+    role: str = None,
+):
+    """
+    Delete rbac permissions of resource.
+    """
+    role_flag = ""
+
+    if role:
+        role_flag = "--role '{}'".format(role)
+
+    cli.invoke(
+        f"role assignment delete --scope '{scope}' --assignee '{assignee}' {role_flag}"
+    )
+
+
+def clean_up_iothub_device_config(
+    hub_name: str,
+    rg: str
+):
+    device_list = []
+    device_list.extend(d["deviceId"] for d in cli.invoke(
+        f"iot hub device-twin list -n {hub_name} -g {rg}"
+    ).as_json())
+
+    config_list = []
+    config_list.extend(c["id"] for c in cli.invoke(
+        f"iot edge deployment list -n {hub_name} -g {rg}"
+    ).as_json())
+
+    config_list.extend(c["id"] for c in cli.invoke(
+        f"iot hub configuration list -n {hub_name} -g {rg}"
+    ).as_json())
+
+    connection_string = cli.invoke(
+        "iot hub connection-string show -n {} -g {} --policy-name {}".format(
+            hub_name, rg, "iothubowner"
+        )
+    ).as_json()["connectionString"]
+
+    if device_list:
+        device = device_list.pop()
```

Comment:
I'm a little confused by the logic here, it looks like we:
- Check if there are devices to delete
- If so:
  - pop the first device
  - delete it using `--login`
  - loop through all other devices
    - delete these using hub name and resource group
    
  The same logic applies below to `config_list`, but could we simplify the logic by just going through `device in device_list` instead of the extra steps for the first item?

Reply:
Hi Ryan, this was the original logic before I touch the code and it's showing green since a factored them out. I was not clear about the history of the --login for first device so I didn't touch this. But after trying out your suggestion I don't see any issue removing the extra first step, therefore will update!😀